### DISCUSSION TITLE
Blade component not loading it's stack when shown using blade conditional

### DISCUSSION THREAD

**Discussion by brentscholl** [livewire/livewire#3131] [7. Bugs (archived)]
### Description
I have a livewire component that has a condition within it. When livewire updates the data for the condition, and shows the updated content based on the condition, it will not load the scripts for the component within the conditional. 

Here is a code example:

```
<div>
@if($client->avatar)
    <div>
        <label>
            Photo
        </label>
        <div>
            <img src="{{ $client->avatarUrl() }}" alt="Client Image">
            <button type="button" wire:click="removeAvatar()">
                Remove
            </button>
        </div>
    </div>
@else
    <div>
        <x-input.image
            wire:model="newAvatar"
            label="Logo"
            :required="false"
            placeholder=""
            multiple
            class=""/>
    </div>
@endif
</div>
```
Here is the x-input.image component
```
@props([
'label' => "",
])

<div wire:ignore
        wire:key="data_{{ $attributes->whereStartsWith('wire:model')->first() }}"
        x-data=""
        x-init="
            FilePond.registerPlugin(FilePondPluginImagePreview);
            FilePond.registerPlugin(FilePondPluginFileValidateType);
            FilePond.registerPlugin(FilePondPluginFileValidateSize);
            FilePond.registerPlugin(FilePondPluginImageResize);

            FilePond.setOptions({
                allowMultiple: {{ isset($attributes['multiple']) ? 'true' : 'false' }},
                server: {
                    process: (fieldName, file, metadata, load, error, progress, abort, transfer, options) => {
                        @this.upload('{{ $attributes->whereStartsWith('wire:model')->first() }}', file, load, error, progress)
                    },
                    revert: (filename, load) => {
                        @this.removeUpload('{{ $attributes->whereStartsWith('wire:model')->first() }}', filename, load)
                    },
                },
            });

            FilePond.create($refs.input, {
                acceptedFileTypes: ['image/png', 'image/jpeg'],
                maxFileSize: '7MB',
                allowImageResize: true,
                imageResizeTargetWidth: '300px',
                imageResizeTargetHeight: '300px',
            });
        "
    >
    <label for="{{ $attributes->whereStartsWith('wire:model')->first() }}">{{ $label }}</label>

    <div>
        <input
            {{ $attributes->whereStartsWith('wire:model') }}
            id="{{ $attributes->whereStartsWith('wire:model')->first() }}"
            type="file" x-ref="input">
    </div>
</div>
@push('styles')
    <link href="https://unpkg.com/filepond/dist/filepond.css" rel="stylesheet">
    <link href="https://unpkg.com/filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css" rel="stylesheet">
@endpush
@push('scripts')
    <script src="https://unpkg.com/filepond-plugin-image-resize/dist/filepond-plugin-image-resize.js"></script>
    <script src="https://unpkg.com/filepond-plugin-file-validate-size/dist/filepond-plugin-file-validate-size.js"></script>
    <script src="https://unpkg.com/filepond-plugin-file-validate-type/dist/filepond-plugin-file-validate-type.js"></script>
    <script src="https://unpkg.com/filepond-plugin-image-preview/dist/filepond-plugin-image-preview.js"></script>
    <script src="https://unpkg.com/filepond/dist/filepond.js"></script>
@endpush
```
So when I update the value of the conditional ie. by clicking the remove button. the component will show, but I get `Uncaught ReferenceError: FilePond is not defined`. When I inspect I see the scripts have not been added to the dom.  If I then manually refresh the page all works fine. 

Is this a bug with livewire or a flaw with my code?

I'm using the latest version of Livewire, Laravel, and Alpine.

**Comment by joshhanley** [comment]
@brentscholl you can't push to styles and scripts stacks in a conditionally rendered item, as Livewire cannot modify the layout file after the first render. You could remove the `@push` tags and instead use `@once` which will include the scripts on the page below the first instance of that component but not after any subsequent ones.

Hope this helps!

```html
@once
<link ...>
<link ...>
<script ...></script>
<script ...></script>
<script ...></script>
@endonce
```

**Comment by brentscholl** [comment]
@joshhanley Thanks for your response! I tried `@once`, but still no luck. 
```
@props([
'label' => "",
])
@once
<link href="https://unpkg.com/filepond/dist/filepond.css" rel="stylesheet">
<link href="https://unpkg.com/filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css" rel="stylesheet">
<script src="https://unpkg.com/filepond-plugin-image-resize/dist/filepond-plugin-image-resize.js"></script>
<script src="https://unpkg.com/filepond-plugin-file-validate-size/dist/filepond-plugin-file-validate-size.js"></script>
<script src="https://unpkg.com/filepond-plugin-file-validate-type/dist/filepond-plugin-file-validate-type.js"></script>
<script src="https://unpkg.com/filepond-plugin-image-preview/dist/filepond-plugin-image-preview.js"></script>
<script src="https://unpkg.com/filepond/dist/filepond.js"></script>
@endonce
<div class="{{ $attributes->get('class') }}"
        wire:ignore
        wire:key="data_{{ $attributes->whereStartsWith('wire:model')->first() }}"
        x-data=""
        x-init="
            FilePond.registerPlugin(FilePondPluginImagePreview);
            FilePond.registerPlugin(FilePondPluginFileValidateType);
            FilePond.registerPlugin(FilePondPluginFileValidateSize);
            FilePond.registerPlugin(FilePondPluginImageResize);

            FilePond.setOptions({
                allowMultiple: {{ isset($attributes['multiple']) ? 'true' : 'false' }},
                server: {
                    process: (fieldName, file, metadata, load, error, progress, abort, transfer, options) => {
                        @this.upload('{{ $attributes->whereStartsWith('wire:model')->first() }}', file, load, error, progress)
                    },
                    revert: (filename, load) => {
                        @this.removeUpload('{{ $attributes->whereStartsWith('wire:model')->first() }}', filename, load)
                    },
                },
            });

            FilePond.create($refs.input, {
                acceptedFileTypes: ['image/png', 'image/jpeg'],
                maxFileSize: '7MB',
                allowImageResize: true,
                imageResizeTargetWidth: '300px',
                imageResizeTargetHeight: '300px',
            });
        "
    >
........
```
When I inspect I can see that the scripts are now at least being added to the page once the conditional changes, but still getting the same error `Uncaught ReferenceError: FilePond is not defined`. I have a feeling it may have something to do with Alpine? This component works fine when shown on the page load, but still not when shown by a conditional update.

**Comment by CraftLogan** [comment]
I currently am having the same problem.

**Comment by joshhanley** [comment]
@brentscholl ah yeah sorry, that's my bad, Livewire will eval new script tags added to the page but only inline scripts.

The only way I can see around it is either ensuring the input is actually loaded on the page on the first render, which will enable the `@push` to work, as they will get pushed into the main app layout before Livewire takes over; or actually adding them into your layout file directly.

If you go with the first option you won't be able to wrap the input in the blade `@if/@else` conditional but you could use Alpine to show/hide each section based on the conditional instead.

Hope this helps!

**Reply by dreanmer** [reply]
woa I was stuck last 3 hours trying to make the `@push` work inside my blade component, and i just realize tha its inside an if/else of the template... is this documented somewhere @joshhanley ? also, there is some way to force it be included (the `@push` from inside the blade component) even if not showing at the first render?

**Reply by joshhanley** [reply]
@dreanmer not sure whether it is documented or not. I've run into that issue a few times myself too.

You'd either have to move the filepond script and style tags to your app layout file, or you could just move them into the Livewire component that contains the input to be toggled into view, so it includes them on the first render, then when the input is rendered, it already has access to the scripts.

Hope this helps!

**Comment by alejandrotrevi** [comment]
@calebporzio What do you think would be the best approach for this kind of situations? 🤔

**Reply by joshhanley** [reply]
@alejandrotrevi we've tried a couple of times to get this running, but it requires breaking changes, so we're hoping to get this into the next major version of Livewire.

Hope this helps!
